src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -38,20 +37,63 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     /// and any consumers observing the results of the task returned from <see cref="ExecuteAsync{TRequestType, TResponseType}(bool, IRequestHandler{TRequestType, TResponseType}, TRequestType, ClientCapabilities, string?, CancellationToken)"/>
     /// will see the results of the handling of the request, whenever it occurred.
     /// </para>
+    /// <para>
+    /// Exceptions in the handling of non-mutating requests are sent back to callers. Exceptions in the processing of
+    /// the queue will close the LSP connection so that the client can reconnect. Exceptions in the handling of mutating
+    /// requests will also close the LSP connection, as at that point the mutated solution is in an unknown state.
+    /// </para>
     /// </remarks>
-    internal partial class RequestExecutionQueue : IDisposable
+    internal partial class RequestExecutionQueue
     {
-        private readonly AsyncQueue<QueueItem> _queue = new AsyncQueue<QueueItem>();
         private readonly ILspSolutionProvider _solutionProvider;
-        private readonly CancellationTokenSource _cancelSource;
+        // The queue and the cancellation token should both be null or non-null at all times
+        private AsyncQueue<QueueItem>? _queue;
+        private CancellationTokenSource? _cancelSource;
+
+        /// <summary>
+        /// Raised when the execution queue has failed, or the solution state its tracking is in an unknown state
+        /// and so the only course of action is to shutdown the server so that the client re-connects and we can
+        /// start over again.
+        /// </summary>
+        /// <remarks>
+        /// Once this event has been fired all currently active and pending work items in the queue will be cancelled.
+        /// </remarks>
+        public event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
 
         public RequestExecutionQueue(ILspSolutionProvider solutionProvider)
         {
             _solutionProvider = solutionProvider;
-            _cancelSource = new CancellationTokenSource();
+        }
+
+        /// <summary>
+        /// Initializes the queue if its not already initialized and starts it processing messages
+        /// </summary>
+        public void Initialize()
+        {
+            // If the queue is already running, do nothing because we don't want to run multiple or lose any queued messages
+            if (_cancelSource == null || _cancelSource.IsCancellationRequested)
```

Comment:
re-initializing when cancellation happens seems pretty confusing.  can we just create a new queue and have the higher layer use that?

Reply:
Yes, the higher level could create a new `RequestExecutionQueue`, sure.